### PR TITLE
build: expand ${ZEPHYR_<module>_MODULE_DIR} in CONF_FILE, OVERLAY_CONFIG

### LINKED ISSUE #41830: CONF_FILE, OVERLAY_CONFIG parsing expands ${ZEPHYR_<whatever>_MODULE_DIR}
**Is your enhancement proposal related to a problem? Please describe.**
Downstream project integrators may be pulling content in from multiple downstream repositories, and may be configuring multiple similar projects. In order to reference config files in another module, integrators must explicitly provide either absolute pathnames, project-dir-relative pathnames, or ZEPHYR_BASE-prefixed pathnames to each file. This solution does not reflect that the mounting points of ZEPHYR_MODULES is arbitrary and is difficult to maintain when an imported west manifest changes the path to a module.

**Describe the solution you'd like**
I would like to see the parsing of the CMake variables CONF_FILE, OVERLAY_CONFIG expand ${ZEPHYR_MYMODULE_MODULE_DIR} substrings into the absolute path to the referenced module.  This enables integrators to reference overlay files in a different module without caring where that module is located in the workspace.

It seems this can be easily implemented where needed similar to the following example:
Replace: 
    `string(REPLACE " " ";" OVERLAY_CONFIG_AS_LIST "${OVERLAY_CONFIG}")`
With:
    `string(CONFIGURE "${OVERLAY_CONFIG}" OVERLAY_CONFIG_EXPANDED)`
    `string(REPLACE " " ";" OVERLAY_CONFIG_AS_LIST "${OVERLAY_CONFIG_EXPANDED}")`

As of Zephyr 2.7.0 it appears this replacement is only needed in `cmake/app/boilerplate.cmake` and `cmake/kconfig.cmake`.

**Describe alternatives you've considered**
I have not considered any other options.

**Additional context**

### PR DESCRIPTION
**Pull request by gregshue**
Support referencing module directories by name in CONF_FILE, and
OVERLAY_CONFIG so that projects can reference overlay files in arbitrary
modules.

Fixes #41830

Signed-off-by: Gregory Shue <gregory.shue@legrand.us>

### CODE REVIEW

**Review by tejlmand — CHANGES_REQUESTED** [review]
Thanks for this proposal.

I see the usefulness in it, but some general thoughts before going into implementation details.

To be consistent between specifying configuration files, then `DTC_OVERLAY_FILE` should also be updated to have similar behavior.

One concern with this approach is that it will be possible to use any CMake variable, for example `-DCONF_FILE=<path>/${FOO}.conf`.
If users start to depend on specific variables inside the build system being defined before the `CONF_FILE` is processed, then we will make it harder for ourselves to refactor our internals, cause suddenly some users might be depending on `FOO` being defined before `CONF_FILE` is used. `FOO` could even be an internal variable.

Of course, some variables could be considered safe, as they will most likely always be available before `CONF_FILE` is used.
For example: `ZEPHYR_<name>_MODULE_DIR` and `BOARD`.
As minimum it should be clearly documented what variables we officially support with this approach.

So please update the documentation to describe this new possibility, together with officially supported variables that can be used.

**Comment by gregshue** [comment]
Thank you. Do you have a doc section in mind for this content?

**Comment by tejlmand** [comment]
> Thank you. Do you have a doc section in mind for this content?

As general description, maybe adding it here:
https://docs.zephyrproject.org/latest/application/index.html#application-configuration-directory
for example as continuation of:
> Zephyr will use configuration files from the application’s configuration directory except for files with an absolute path provided by the arguments described earlier, for example CONF_FILE, OVERLAY_CONFIG, and DTC_OVERLAY_FILE.


For `CONF_FILE` I would suggest we also adding this knowledge here:
https://docs.zephyrproject.org/latest/guides/build/kconfig/setting.html#the-initial-configuration
> All configuration files will be taken from the application’s configuration directory except for files with an absolute path that are given with the CONF_FILE argument.

with reference to variables that are officially supported with this method.

For devicetree overlays, here:
https://docs.zephyrproject.org/latest/guides/dts/howtos.html#set-devicetree-overlays

**Review by tejlmand — CHANGES_REQUESTED** [review]
Thanks for updating the docs.

Some comments, to ensure it become more precise.

File: doc/guides/dts/howtos.rst [hunk 1]
```diff
@@ -202,7 +202,21 @@ Set devicetree overlays
...
+
+   set(DTC_OVERLAY_FILE "\${ZEPHYR_MODULENAME_MODULE_DIR}/<module-relative-pathname>")
+
+   find_package(Zephyr)
+
+or from the shell as:
+
+.. code-block:: shell
+
+   cmake -DDTC_OVERLAY_FILE="\${ZEPHYR_MODULENAME_MODULE_DIR}/<module-relative-pathname>"
```
Comment (tejlmand) [inline doc/guides/dts/howtos.rst:219]: Related to the comment of the rendered output.

How about just a short sentence on this, for example like:
```suggestion
files, they are included in that order by the C preprocessor. A file in a Zephyr
module can be referred by escaping the Zephyr module dir variable like `\${ZEPHYR_<module>_MODULE_DIR}/<path-to>/dts.overlay` when setting the DTC_OVERLAY_FILE variable.
```

File: doc/application/index.rst [hunk 2]
```diff
@@ -478,7 +478,20 @@ Application Configuration Directory
...
+
+   set(CONF_FILE "\${ZEPHYR_MODULENAME_MODULE_DIR}/<module-relative-pathname>")
+
+   find_package(Zephyr)
+
+or from the shell as:
+
+.. code-block:: shell
+
+   cmake -DCONF_FILE="\${ZEPHYR_MODULENAME_MODULE_DIR}/<module-relative-pathname>"
```
Comment (tejlmand) [inline doc/application/index.rst:494]: Looking at the rendered output, I feel this particular feature fills to much, compared to the description of the other parts:
image

How about just a short mentioning of this possibility, for example:
```suggestion
``DTC_OVERLAY_FILE``. A file in a Zephyr module can be referred by escaping the Zephyr module dir variable like this ``\${ZEPHYR_<module>_MODULE_DIR}/<path-to>/<file>`` when setting the any of said variables.
```

File: tests/cmake/overlays/var_expansions/my_extra_module/zephyr/my_extra_module-board.overlay [hunk 3]
```diff
@@ -0,0 +1,10 @@
+/*
+ * Copyright (c) 2022 Legrand, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/ {
+        /* This file is intentionally empty */
+};
+
```
Comment (tejlmand) [inline tests/cmake/overlays/var_expansions/my_extra_module/zephyr/my_extra_module-board.overlay:10]:
```suggestion
```

File: tests/cmake/overlays/var_expansions/my_module/zephyr/my_module-board.overlay [hunk 4]
```diff
@@ -0,0 +1,10 @@
+/*
+ * Copyright (c) 2022 Legrand, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/ {
+        /* This file is intentionally empty */
+};
+
```
Comment (tejlmand) [inline tests/cmake/overlays/var_expansions/my_module/zephyr/my_module-board.overlay:10]:
```suggestion
```

**Review by tejlmand — CHANGES_REQUESTED** [review]
Thanks for the update.

I took a look at the rendered output, and have a small proposal.

File: doc/application/index.rst [hunk 5]
```diff
@@ -478,7 +478,7 @@ Application Configuration Directory
 Zephyr will use configuration files from the application's configuration
 directory except for files with an absolute path provided by the arguments
 described earlier, for example ``CONF_FILE``, ``OVERLAY_CONFIG``, and
-``DTC_OVERLAY_FILE``.
+``DTC_OVERLAY_FILE``. A file in a Zephyr module can be referred by escaping the Zephyr module dir variable like this ``\${ZEPHYR_<module>_MODULE_DIR}/<path-to>/<file>`` when setting the any of said variables.
```
Comment (tejlmand) [inline doc/application/index.rst:481]: Whoops, seems my suggestion had a stray `the`

```suggestion
``DTC_OVERLAY_FILE``. A file in a Zephyr module can be referred by escaping the Zephyr module dir variable like this ``\${ZEPHYR_<module>_MODULE_DIR}/<path-to>/<file>`` when setting any of said variables.
```

**Review by tejlmand — DISMISSED** [review]
lgtm, thanks for this.

Found a stray `the` to be removed before merging, but approving as everything else is fine.

**Review by tejlmand — APPROVED** [review]
lgtm - re-approved

**Review by de-nordic — APPROVED** [review]

**Comment by tejlmand** [comment]
@gregshue seems this has gone stale.
If you still would like it in, please re-open and rebase, then i'll re-approve

**Comment by dleach02** [comment]
@nashif, are you able to reopen this? The button is greyed out for me

**Comment by gregshue** [comment]
@nashif Are you able to reopen this?  I'm not given the option.

**Comment by gregshue** [comment]
@tejlmand @dleach02 @nashif Since we are having trouble re-opening this PR, I created a new one with the same changes (#47638)

**Pull request closed without merging**